File: 01_Data_Structures/Stack/LC_155_Min_Stack.cpp
/**
 * @file LC_155_Min_Stack.cpp
 * @author BalaSeethaRamanjaneyulu
 * @brief Solution for LeetCode 155: Min Stack
 *
 * Problem: Design a stack that supports push, pop, top, and retrieving 
 * the minimum element in constant time.
 *
 * Platform: LeetCode #155
 * Approach: Two Stacks (Data Stack + Min Stack)
 * Time:     O(1) for all operations
 * Space:    O(N)
 * Link:     https://leetcode.com/problems/min-stack/
 */

#include <iostream>
#include <stack>
#include <algorithm>

using namespace std;

/**
 * @brief A stack that supports O(1) minimum retrieval.
 */
class MinStack {
private:
    stack<int> s;      // Main data stack
    stack<int> min_s;  // Auxiliary stack to track minimums

public:
    MinStack() {
    }
    
    /**
     * @brief Pushes element val onto stack.
     */
    void push(int val) {
        s.push(val);
        // If min stack is empty or val is new minimum, push to min stack
        if (min_s.empty() || val <= min_s.top()) {
            min_s.push(val);
        }
    }
    
    /**
     * @brief Removes the element on the top of the stack.
     */
    void pop() {
        // If the element being popped is the current minimum, pop from min stack too
        if (s.top() == min_s.top()) {
            min_s.pop();
        }
        s.pop();
    }
    
    /**
     * @brief Gets the top element of the stack.
     */
    int top() {
        return s.top();
    }
    
    /**
     * @brief Retrieves the minimum element in the stack.
     */
    int getMin() {
        return min_s.top();
    }
};

// ─── Driver ──────────────────────────────────────────────────────────────────
int main() {
    MinStack* obj = new MinStack();
    obj->push(-2);
    obj->push(0);
    obj->push(-3);
    
    cout << "Min after push -3: " << obj->getMin() << endl; // Should be -3
    
    obj->pop();
    cout << "Top after pop: " << obj->top() << endl;       // Should be 0
    cout << "Min after pop: " << obj->getMin() << endl;    // Should be -2
    
    delete obj;
    return 0;
}

File: 01_Data_Structures/Stack/LC_155_Min_Stack.java
/**
 * @file LC_155_Min_Stack.java
 * @author BalaSeethaRamanjaneyulu
 * @brief Java Solution for LeetCode 155: Min Stack
 */

import java.util.Stack;

public class LC_155_Min_Stack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public LC_155_Min_Stack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    /**
     * Pushes val onto the stack.
     * 
     * @param val The value to push.
     */
    public void push(int val) {
        stack.push(val);
        // If minStack is empty or val <= current minimum, push to minStack
        if (minStack.isEmpty() || val <= minStack.peek()) {
            minStack.push(val);
        }
    }

    /**
     * Removes the top element.
     */
    public void pop() {
        // Use .equals() for object comparison
        if (stack.peek().equals(minStack.peek())) {
            minStack.pop();
        }
        stack.pop();
    }

    /**
     * Gets the top element.
     * 
     * @return int The top element.
     */
    public int top() {
        return stack.peek();
    }

    /**
     * Gets the minimum element.
     * 
     * @return int The current minimum.
     */
    public int getMin() {
        return minStack.peek();
    }

    // ─── Driver ──────────────────────────────────────────────────────────────
    public static void main(String[] args) {
        LC_155_Min_Stack obj = new LC_155_Min_Stack();
        obj.push(-2);
        obj.push(0);
        obj.push(-3);
        System.out.println("Min: " + obj.getMin()); // -3
        obj.pop();
        System.out.println("Top: " + obj.top());    // 0
        System.out.println("Min: " + obj.getMin()); // -2
    }
}
